docs/proposals/role-entry-points.md

```diff
@@ -0,0 +1,264 @@
+# Entry-point based roles
+
+## We have a problem
+
+Each role consists of a set of tasks and variables read/set in some of the tasks.
+Some of the tasks can store their results into registered variables,
+e.g. store stdout of a docker pull command.
+Some of the tasks may be run only under certain condition that are generated
+outside of a role, e.g. pull images only if a deployment is containerized.
+Some of the tasks are run only if a certain variable is defined,
+e.g. add additional Kubelet arguments into the `node-config.yml`
+if `kubelet_additional_args` inventory variable is defined.
+
+Some roles work with a lot of variables that from a point of a role are uninvited citizens.
+Some of the variables are set inside a role, some of them are set outside of a role.
+The variables outside of a role are hard to track, hard to determine their value or
+a condition under which the variables are defined. These unknown facts make a role harder
+to maintain, harder to make a role closed with respect to variables and to make
+a role practically modular.
+
+## Proposal
+
+Make a role that conforms to the following conditions:
+* none of the variables read/set inside a role are global
+* all variables inside a role are `r_` (`r` standing for a role) and/or `l_`(`l` standing for a local) prefixed
+* all values are propagated to a role through `r_` prefixed variables only
+* all values that are set inside a role are read outside of a role through `r_` prefixed variables only
+* all available `r_` prefixed variables are documented and checked before a main role task is run
+* a table of `r_` prefixed variables in README.md is automatically generated
+
+### Terminilogy
+
+**Role variables**: All `r_` prefixed variables read/set inside a role
+
+**Local variables**: All `l_` prefixed variables set inside a role
+
+### Role properties
+
+**Solid API**: all variables that a role caller can set need to be well defined.
+It means each such variable must have the following information attached:
+* name
+* description
+* default value (if the variable is defined)
+* set of acceptable values
+* if it is required or if it is optional
+
+Well defined API allows automatically generated documentation and variable checks.
+
+**Free of outer variables**: only variables consisting the role API or defined locally are allowed.
```

Comment:
Some non-role variables has to be set everytime, like the `openshift_version`. So either it is set inside an inventory or it is set by `openshift_version` role. Other non-role variables can be undefined unless set in an inventory file like `openshift_node_kubelet_args`. If the variable is not set, I should not set corresponding `r_openshift_node_kubelet_args` to anything when invoking a role. If the variable is set, I should set `r_openshift_node_kubelet_args: "{{ openshift_node_kubelet_args }}"`

The question is how can I do that in the following form for both cases:
```yaml
- include_role:
    name: openshift_node
  vars:
    r_openshift_node_kubelet_args: "{{ openshift_node_kubelet_args }}"
```

without using the `default({})` construct? I.e.

```yaml
- include_role:
    name: openshift_node
  vars:
    r_openshift_node_kubelet_args: "{{ openshift_node_kubelet_args | default({}) }}"
```

I see only one way to do it and it is:
```yaml
- set_fact:
    r_openshift_node_kubelet_args: "{{ openshift_node_kubelet_args }}"
  when: openshift_node_kubelet_args is defined

- include_role:
    name: openshift_node
  vars:
    ...
```

Or, we need to make sure that all inventory variables are always set to some default value even if they are not specified in an inventory file.